### PR TITLE
[Feature] Categorical encoding for action space

### PR DESCRIPTION
**Pull request by artkorenev**
## Description

Added alternative to one-hot encoding for action spaces with categorical features.

## Motivation and Context

Implementing feature #538.

## Types of changes

What types of changes does your code introduce? Remove all that do not apply:

- [x] New feature (non-breaking change which adds core functionality)

## Checklist
* [x]  I have read the [CONTRIBUTION](https://github.com/pytorch/rl/blob/main/CONTRIBUTING.md) guide (**required**)
* [ ]  My change requires a change to the documentation.
* [x]  I have updated the tests accordingly (_required for a bug fix or a new feature_).
* [ ]  I have updated the documentation accordingly.

Checklist from #538:
* [x]  Create a `DiscreteTensorSpec` [similar to `MultOneHotDiscreteTensorSpec`](https://github.com/pytorch/rl/blob/417817abc12240cb8c93d767445fdec5f6c7050c/torchrl/data/tensor_specs.py#L909) + tests
* [x]  Change QValueHook to make it possible to use [a "categorical" space (or similar) instead of "one_hot"](https://github.com/pytorch/rl/blob/497ad8f6304403f358f093f289d4c94ad2ce5878/torchrl/modules/tensordict_module/actors.py#L259)
* [x]  Do the same with [`DistributionalQValueHook`](https://github.com/pytorch/rl/blob/497ad8f6304403f358f093f289d4c94ad2ce5878/torchrl/modules/tensordict_module/actors.py#L357)
* [x]  [Adapt the DQNLoss](https://github.com/pytorch/rl/blob/497ad8f6304403f358f093f289d4c94ad2ce5878/torchrl/objectives/costs/dqn.py#L91-L104) to these changes
* [x]  [Adapt the DistributionalDQNLoss](https://github.com/pytorch/rl/blob/497ad8f6304403f358f093f289d4c94ad2ce5878/torchrl/objectives/costs/dqn.py#L129) to these changes
* [x]  Change the gym specs reader [here](https://github.com/pytorch/rl/blob/497ad8f6304403f358f093f289d4c94ad2ce5878/torchrl/envs/libs/gym.py#L58) to make it possible to choose one spec type or the other. One way to go about this would be to have a global variable that controls which categorical encoding must be used.

Current implementation seems to couple different parts of TorchRL quite tightly as there is quite a few places where it is assumed that the action space is one-hot. Perhaps, a more generic action-space object that encapsulates manipulation with values and that is shared across losses, actors, hooks, etc. (as a `nn.Module` maybe) might be a worth investment in future.

The mode is turned on by `export CATEGORICAL_ACTION_ENCODING=True` command, and it is turned off by default. Another alternative could be to provide it through the config-file, however this, again, would require a bit more generic solution to configure module that will be passed to all the necessary modules.

### CODE REVIEW

File: torchrl/_utils.py [hunk 1]
```diff
@@ -117,3 +118,24 @@ def prod(sequence):
         return math.prod(sequence)
     else:
         return int(np.prod(sequence))
+
+
+def get_binary_env_var(key):
```
Comment (vmoens) [inline torchrl/_utils.py:123]: codecov complains about this being half tested, can we fix that?

File: torchrl/data/tensor_specs.py [hunk 2]
```diff
@@ -37,27 +36,20 @@
...
     "MultOneHotDiscreteTensorSpec",
+    "DiscreteTensorSpec",
     "CompositeSpec",
 ]
 
+from torchrl._utils import get_binary_env_var
 from torchrl.data.tensordict.tensordict import TensorDictBase, TensorDict
 
-_CHECK_IMAGES = os.environ.get("CHECK_IMAGES", False)
+_CHECK_IMAGES = get_binary_env_var("CHECK_IMAGES")
```
Comment (vmoens) [inline torchrl/data/tensor_specs.py:46]: thanks for the fix

File: torchrl/data/tensor_specs.py [hunk 3]
```diff
@@ -997,6 +992,72 @@ def _project(self, val: torch.Tensor) -> torch.Tensor:
         return torch.cat([super()._project(_val) for _val in vals], -1)
 
 
+class DiscreteTensorSpec(TensorSpec):
+    """A discrete tensor spec.
+
+    An alternative to OneHotTensorSpec for categorical variables in TorchRL. Instead of
+    using multiplication, categorical variables perform indexing which can speed up
+    computation and reduce memory cost for large categorical variables.
+        >>> batch, size = 3, 4
```
Comment (vmoens) [inline torchrl/data/tensor_specs.py:1001]: Can we put that under an `Example:`?

File: torchrl/data/tensor_specs.py [hunk 4]
```diff
@@ -997,6 +992,72 @@ def _project(self, val: torch.Tensor) -> torch.Tensor:
...
+    shape: torch.Size
+    space: DiscreteBox
+    device: torch.device = torch.device("cpu")
+    dtype: torch.dtype = torch.float
+    domain: str = ""
+
+    def __init__(
+        self,
+        n: int,
+        shape: Optional[torch.Size] = torch.Size((1,)),
```
Comment (vmoens) [inline torchrl/data/tensor_specs.py:1026]: IIRC we once had a linting issue with this (that seems to have disappeared now).
This is why we usually have a `None` and if `None` then shape is replaced by `Size([1])`.

File: torchrl/data/tensor_specs.py [hunk 5]
```diff
@@ -997,6 +992,72 @@ def _project(self, val: torch.Tensor) -> torch.Tensor:
...
+
+    def rand(self, shape=None) -> torch.Tensor:
+        if shape is None:
+            shape = torch.Size([])
+        return (torch.rand(*shape, *self.shape, device=self.device) * self.space.n).to(
+            torch.long
+        )
+
+    def _project(self, val: torch.Tensor) -> torch.Tensor:
+        return torch.round(val).clamp_(min=0, max=self.space.n - 1)
```
Comment (vmoens) [inline torchrl/data/tensor_specs.py:1042]: We should test the project method for this class

File: torchrl/modules/tensordict_module/actors.py [hunk 6]
```diff
@@ -294,6 +306,21 @@ def _mult_one_hot(self, value: torch.Tensor, support: torch.Tensor) -> torch.Ten
...
+        values: torch.Tensor, action: torch.Tensor
+    ) -> torch.Tensor:
+        return (action * values).sum(-1, True)
+
+    @staticmethod
+    def _categorical_action_value(
+        values: torch.Tensor, action: torch.Tensor
+    ) -> torch.Tensor:
+        if len(values.shape) == 1:
+            return values[action].unsqueeze(-1)
```
Comment (vmoens) [inline torchrl/modules/tensordict_module/actors.py:320]: this seems to be untested

File: torchrl/modules/tensordict_module/actors.py [hunk 7]
```diff
@@ -352,18 +379,23 @@ def __init__(
...
         self.var_nums = var_nums
-        self.fun_dict = {
+        self.action_func_mapping = {
             "one_hot": self._one_hot,
             "mult_one_hot": self._mult_one_hot,
             "binary": self._binary,
+            "categorical": self._categorical,
         }
+        if action_space not in self.action_func_mapping:
+            raise ValueError(
```
Comment (vmoens) [inline torchrl/modules/tensordict_module/actors.py:389]: this is not covered by the tests

File: torchrl/modules/tensordict_module/actors.py [hunk 8]
```diff
@@ -401,6 +433,10 @@ def _mult_one_hot(self, value: torch.Tensor, support: torch.Tensor) -> torch.Ten
             -1,
         )
 
+    def _categorical(self, value: torch.Tensor, support: torch.Tensor) -> torch.Tensor:
+        value = value = self._support_expected(value, support)
+        return torch.argmax(value, dim=-1).to(torch.long)
```
Comment (vmoens) [inline torchrl/modules/tensordict_module/actors.py:438]: this seems to be untested

File: torchrl/objectives/dqn.py [hunk 9]
```diff
@@ -95,8 +96,15 @@ def forward(self, input_tensordict: TensorDictBase) -> TensorDict:
             buffers=self.value_network_buffers,
         )
 
+        action = tensordict.get("action")
         pred_val = td_copy.get("action_value")
-        pred_val_index = (pred_val * action).sum(-1)
+
+        if _CATEGORICAL_ACTION_ENCODING:
+            batch_size = action.size(0)
+            pred_val_index = pred_val[range(batch_size), action.squeeze(-1)]
```
Comment (vmoens) [inline torchrl/objectives/dqn.py:104]: I guess we'd like to be able to work with batch sizes that are more than unidimensional.

We can use `gather` (I think)
```python
x = torch.randn(3, 4, 5)
idx = torch.randint(5, (3, 4))
x.gather(-1, idx)
```

This will index `x` along the last dimension.

Comment (vmoens) [reply torchrl/objectives/dqn.py:104]: this is not covered by the tests (it should be done in `test_costs.py` I believe).

File: torchrl/objectives/dqn.py [hunk 10]
```diff
@@ -95,8 +96,15 @@ def forward(self, input_tensordict: TensorDictBase) -> TensorDict:
             buffers=self.value_network_buffers,
         )
 
+        action = tensordict.get("action")
         pred_val = td_copy.get("action_value")
-        pred_val_index = (pred_val * action).sum(-1)
+
+        if _CATEGORICAL_ACTION_ENCODING:
```
Comment (vmoens) [inline torchrl/objectives/dqn.py:102]: This is a bit tricky

What if someone explicitly builds a policy that uses categorical actions, but does not set the env variable?

My view of the env variable is more along the line of "a way of easily change a whole training script from categorical to one hot and vice-versa".

But the reversed dependency should not hold, and the environment variable should only be checked when building the network.

One option could be to infer the type of action by looking at the `value_network`: if it has registered the action spec, then we can use it to infer the type of action we will see. Otherwise, we must ask to the user to tell us (via an arg in the constructor) what kind of action we should be expecting.

Open to other suggestions obviously :)

File: torchrl/objectives/dqn.py [hunk 11]
```diff
@@ -198,9 +206,13 @@ def forward(self, input_tensordict: TensorDictBase) -> TensorDict:
             buffers=self.value_network_buffers,
         )  # Log probabilities log p(s_t, ·; θonline)
         action_log_softmax = td_clone.get("action_value")
-        action_expand = action.unsqueeze(-2).expand_as(action_log_softmax)
-        log_ps_a = action_log_softmax.masked_select(action_expand.to(torch.bool))
-        log_ps_a = log_ps_a.view(batch_size, atoms)  # log p(s_t, a_t; θonline)
+
+        if _CATEGORICAL_ACTION_ENCODING:
+            log_ps_a = action_log_softmax[range(batch_size), :, action.squeeze(-1)]
```
Comment (vmoens) [inline torchrl/objectives/dqn.py:211]: same comments as above regarding indexing, usage of env variable and test coverage

File: torchrl/trainers/helpers/models.py [hunk 12]
```diff
@@ -178,6 +181,11 @@ def make_dqn_actor(
     out_features = env_specs["action_spec"].shape[0]
     actor_class = QValueActor
     actor_kwargs = {}
+
+    if _CATEGORICAL_ACTION_ENCODING:
+        actor_kwargs.update({"action_space": "categorical"})
+        out_features = env_specs["action_spec"].space.n
```
Comment (vmoens) [inline torchrl/trainers/helpers/models.py:187]: perhaps we could have an arg in the function that indicates if categorical has to be used.
If default (e.g. `None`) then it falls back on `_CATEGORICAL_ACTION_ENCODING`.

**Review by vmoens — COMMENTED** [review]
We'd like to add the new classes to the docs (this is not done automatically atm).
Just add them in docs/source/data.

File: torchrl/objectives/dqn.py (hunk 10 above)
Comment (artkorenev) [reply torchrl/objectives/dqn.py:102]: > What if someone explicitly builds a policy that uses categorical actions, but does not set the env variable?

This is exactly what I meant when mentioned that the current solution couples things a bit too much.

Since actions are passed as pure tensors (even though, wrapped in TensorDict) it is either assumed or configured separately for each class what to do with `action` tensors.

Now I can see two options what we can do here:

1. Whenever we deal with actions, we either pass `action_space` in the constructor or derive it from some other entity (e.g. `value_network` as you proposed). However here we rely on careful configuration of all modules together (in theory, somebody could explicitly configure `value_network` incompatible with an environment). Another drawback is that we "spreading out" the logic for each case across the code base which can make code a harder to understand (I think in case of one-hot/categorical it is fine, but for future approaches it can blow up).
2. We can introduce something like `ActionSpec` which would be an extended version of existing specs, though purely for action processing (deriving actions from `value_network`, etc.). We instantiate the `ActionSpec` once and pass its instance everywhere we need encapsulating work with actions. Although, it is hard for me to estimate what interface `ActionSpec` needs in order to cover all possible cases (I mean globally for all RL cases). And this also would require larger scale refactoring.

tldr; I think 1st option seems better choice for this situation and somewhat follows current code base practices (e.g. `QValueHook`)

File: torchrl/_utils.py [hunk 13]
```diff
@@ -117,3 +118,24 @@ def prod(sequence):
...
+    """Parses and returns the binary enironment variable value.
+
+    If not present in environment, it is considered `False`.
+
+    Args:
+        key (str): name of the environment variable.
+    """
+    val = os.environ.get(key, False)
+    if val in ("0", "False", False):
```
[the remark this answers is not in the document]
Comment (artkorenev) [reply torchrl/_utils.py]: I changed it to `"False"` by default. Environment variables cannot be set to something other than `str`.

File: torchrl/_utils.py (hunk 1 above)
Comment (artkorenev) [reply torchrl/_utils.py:123]: Added tests in order to cover every case.

File: torchrl/data/tensor_specs.py (hunk 3 above)
Comment (artkorenev) [reply torchrl/data/tensor_specs.py:1003]: Fixed, thx

File: torchrl/data/tensor_specs.py (hunk 4 above)
Comment (artkorenev) [reply torchrl/data/tensor_specs.py]: Noted, changed to that

File: torchrl/data/tensor_specs.py (hunk 5 above)
Comment (artkorenev) [reply torchrl/data/tensor_specs.py]: Added tests for `project`, thx

File: torchrl/modules/tensordict_module/actors.py (hunk 6 above)
Comment (artkorenev) [reply torchrl/modules/tensordict_module/actors.py:320]: Added tests for Q value hooks now

File: torchrl/modules/tensordict_module/actors.py (hunk 7 above)
Comment (artkorenev) [reply torchrl/modules/tensordict_module/actors.py:389]: Fixed with tests

File: torchrl/modules/tensordict_module/actors.py (hunk 8 above)
Comment (artkorenev) [reply torchrl/modules/tensordict_module/actors.py:438]: This should be covered now as well

File: torchrl/objectives/dqn.py (hunk 9 above)
Comment (artkorenev) [reply torchrl/objectives/dqn.py]: Reworked indexing here, thx

File: torchrl/objectives/dqn.py (hunk 11 above)
Comment (artkorenev) [reply torchrl/objectives/dqn.py]: Reworked this as well with a few tweaks since we work with atoms here.

File: torchrl/trainers/helpers/models.py (hunk 12 above)
Comment (artkorenev) [reply torchrl/trainers/helpers/models.py:191]: I changed it here to be inferred from action spec that is set during the env setup.

File: torchrl/objectives/dqn.py (hunk 9 above)
Comment (artkorenev) [reply torchrl/objectives/dqn.py]: Also added tests so should be covered by coverage by now

**Review by vmoens — APPROVED** [review]
Great! Amazing work, if fits in the library very well now. I think we're there. 

A couple of things before we merge this
- Can you have a look at the few comments I left?
- Can you have a look at the coverage report (some lines are not covered -- don't worry about things that are not covered but are in the `test` directory).

File: test/test_actors.py [hunk 14]
```diff
@@ -0,0 +1,126 @@
+import pytest
+import torch
+from torchrl.modules.tensordict_module.actors import (
+    QValueHook,
+    DistributionalQValueHook,
+)
+
+
+def test_qvalue_hook_wrong_action_space():
+    with pytest.raises(ValueError):
```
[the remark this answers is not in the document]
Comment (artkorenev) [reply test/test_actors.py]: Added a message check (had to make it short since it relies on order of the items in dict)

File: test/test_actors.py [hunk 15]
```diff
@@ -0,0 +1,126 @@
...
+        QValueHook(action_space="wrong_value")
+
+
+def test_distributional_qvalue_hook_wrong_action_space():
+    with pytest.raises(ValueError):
```
[the remark this answers is not in the document]
Comment (artkorenev) [reply test/test_actors.py]: Fixed as in `test_qvalue_hook_wrong_action_space`

File: test/test_actors.py [hunk 16]
```diff
@@ -0,0 +1,126 @@
+import pytest
+import torch
+from torchrl.modules.tensordict_module.actors import (
+    QValueHook,
+    DistributionalQValueHook,
+)
+
+
+def test_qvalue_hook_wrong_action_space():
```
[the remark this answers is not in the document]
Comment (artkorenev) [reply test/test_actors.py]: > could we put all those `test_qvalue` under a `TestQValue` class?
Fixed, thanks

**Comment by artkorenev** [comment]
Awesome! Thank you for the review!
It seems like the coverage is now fixed completely.
